File: pipeline/pipeline/registry.py
"""Hot-reloadable model registry, mirrored from server/internal/models.

Rows are immutable and versioned. The cache never evicts a
``(provider_slug, model_slug, version)`` pin. Polling ``model_registry_state``
only teaches this process the current defaults.

Two rules keep a resolved model from drifting away from the one that was priced:

* A cache miss never falls back to the current default.
* :func:`resolve_pinned` requires an exact pin for **every** surface. Nothing
  that can bill, and nothing that writes a vector, is allowed to pick a model
  for itself; the caller that reserved the spend, enqueued the job or created
  the workspace already chose one. ``registry.default`` remains for the
  caller whose job it is to choose (ingest enqueue in the gateway).
"""

from __future__ import annotations

import logging
import threading
import time
from contextvars import ContextVar
from dataclasses import dataclass, field
from typing import Any

from .generated import Surface
from .store import db

log = logging.getLogger("evo.registry")

POLL_INTERVAL = 600.0

AUTH_PLATFORM = "platform"
AUTH_USER_KEY = "user_key"
AUTH_PLATFORM_OR_USER = "platform_or_user"


def join_model_label(provider_name: str, model_name: str) -> str:
    provider = (provider_name or "").strip()
    model = (model_name or "").strip()
    if not provider:
        return model
    if not model:
        return provider
    return f"{provider} {model}"


@dataclass(frozen=True)
class ModelConfig:
    version: int
    provider_name: str
    model_name: str
    provider_slug: str
    model_slug: str
    platform_enabled: bool = True
    byok_enabled: bool = False
    params: dict[str, Any] = field(default_factory=dict)
    surfaces: tuple[Surface, ...] = ()
    thinking_levels: tuple[str, ...] = ()
    default_thinking: str = ""
    micros_per_input_token: int = 0
    micros_per_output_token: int = 0
    micros_per_cached_input_token: int = 0
    enabled: bool = True
    is_default_for: tuple[Surface, ...] = ()
    context_window_tokens: int = 0

    @property
    def auth_mode(self) -> str:
        if self.platform_enabled and self.byok_enabled:
            return AUTH_PLATFORM_OR_USER
        if self.byok_enabled:
            return AUTH_USER_KEY
        return AUTH_PLATFORM

    @property
    def litellm_model_id(self) -> str:
        return self.model_slug

    def resolve_thinking(self, stored: str) -> str:
        if not stored:
            if self.default_thinking in self.thinking_levels:
                return self.default_thinking
            raise RegistryError(
                f"model {self.provider_slug}/{self.model_slug} has no valid "
                "default thinking"
            )
        if stored in self.thinking_levels:
            return stored
        raise RegistryError(
            f"model {self.provider_slug}/{self.model_slug} does not support "
            f"thinking {stored!r}"
        )

    @property
    def pin(self) -> tuple[str, str, int]:
        return (self.provider_slug, self.model_slug, self.version)

    def allows(self, surface: Surface) -> bool:
        return surface in self.surfaces

    @property
    def embedding_dim(self) -> int:
        """Width this model emits, which selects the vector table it writes to.

        Required on every embedding row by a check constraint in the migration,
        so a missing value means the row was written around the schema.
        """
        raw = self.params.get("dimensions")
        if raw is None:
            raise RegistryError(
                f"embedding model {self.provider_slug}/{self.model_slug} "
                f"v{self.version} declares no "
                "dimensions"
            )
        return int(raw)

    def temperature(self, fallback: float = 0.3) -> float:
        raw = self.params.get("temperature", fallback)
        try:
            return float(raw)
        except (TypeError, ValueError):
            return fallback

    def api_mode(self) -> str:
        return "chat_completion"

    @property
    def display_name(self) -> str:
        return join_model_label(self.provider_name, self.model_name)


class RegistryError(RuntimeError):
    """An exact provider/model/version pin could not be loaded."""


@dataclass
class JobPins:
    ingest: ModelConfig | None = None
    embedding: ModelConfig | None = None
    vision: ModelConfig | None = None


_job_pins: ContextVar[JobPins | None] = ContextVar("job_pins", default=None)

_select_cols = """
                    SELECT version, provider_name, model_name, provider_slug,
                           model_slug, platform_enabled, byok_enabled,
                           params, surfaces, thinking_levels, default_thinking,
                           micros_per_input_token, micros_per_output_token,
                           enabled, is_default_for, context_window_tokens,
                           micros_per_cached_input_token
                      FROM model_configs
"""


@dataclass
class RequestLLM:
    user_id: str = ""
    paid_by: str = ""
    user_api_key: str = ""
    thinking: str = ""


_request_llm: ContextVar[RequestLLM | None] = ContextVar("request_llm", default=None)


def bind_request_llm(
    user_id: str | None = None,
    paid_by: str | None = None,
    thinking: str | None = None,
    *,
    user_api_key: str | None = None,
) -> None:
    _request_llm.set(
        RequestLLM(
            user_id=(user_id or "").strip(),
            paid_by=(paid_by or "").strip(),
            user_api_key=(user_api_key or "").strip(),
            thinking=(thinking or "").strip(),
        )
    )


def current_request_llm() -> RequestLLM:
    return _request_llm.get() or RequestLLM()


def set_job_pins(pins: JobPins | None) -> None:
    _job_pins.set(pins)


def current_job_pins() -> JobPins | None:
    return _job_pins.get()


class Registry:
    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._by_pin: dict[tuple[str, str, int], ModelConfig] = {}
        self._current: dict[str, tuple[str, str, int]] = {}
        self._rev: int = -1
        self._started = False

    def start(self) -> None:
        # There is deliberately no process-start snapshot of the embedding or
        # vision default any more. Freezing them was how this process used to
        # avoid mixing vector spaces, which made the model a query embedded with
        # a function of when the container last booted. Embedding now comes from
        # the workspace embedding pin and vision from the job pin, so there
        # is nothing left for a poll to corrupt.
        self.refresh()
        with self._lock:
            self._started = True

    def refresh(self) -> None:
        try:
            with db.connect() as conn, conn.cursor() as cur:
                cur.execute("SELECT version FROM model_registry_state WHERE id = true")
                row = cur.fetchone()
                rev = int(row[0]) if row else 0
                with self._lock:
                    if self._rev == rev and self._current:
                        return
                cur.execute(_select_cols)
                rows = cur.fetchall()
        except Exception:  # noqa: BLE001 - retain the last successfully loaded catalog
            if not self._started:
                log.warning("model registry unavailable; no catalog loaded")
            return

        by_pin: dict[tuple[str, str, int], ModelConfig] = {}
        current: dict[str, tuple[str, str, int]] = {}
        for row in rows:
            spec = _from_row(row)
            by_pin[spec.pin] = spec
            if not spec.enabled:
                continue
            for surface in spec.is_default_for:
                prev = current.get(surface)
                if prev is None or spec.version >= prev[2]:
                    current[surface] = spec.pin

        with self._lock:
            self._by_pin.update(by_pin)
            self._current.update(current)
            self._rev = rev

    def get(self, provider_slug: str, model_slug: str, version: int) -> ModelConfig:
        if not provider_slug or not model_slug or version <= 0:
            raise RegistryError(
                f"invalid pin {provider_slug!r}/{model_slug!r} v{version}"
            )
        pin = (provider_slug, model_slug, version)
        with self._lock:
            cached = self._by_pin.get(pin)
        if cached is not None:
            return cached
        spec = self._load(provider_slug, model_slug, version)
        with self._lock:
            self._by_pin[pin] = spec
        return spec

    def _load(self, provider_slug: str, model_slug: str, version: int) -> ModelConfig:
        with db.connect() as conn, conn.cursor() as cur:
            cur.execute(
                _select_cols
                + " WHERE provider_slug=%s AND model_slug=%s AND version=%s",
                (provider_slug, model_slug, version),
            )
            row = cur.fetchone()
        if not row:
            raise RegistryError(
                f"model config not found: {provider_slug}/{model_slug} v{version}"
            )
        return _from_row(row)

    def default_pin(self, surface: Surface) -> tuple[str, str, int]:
        with self._lock:
            pin = self._current.get(surface)
        if not pin:
            raise RegistryError(f"no default for {surface}")
        return pin

    def default(self, surface: Surface) -> ModelConfig:
        provider_slug, model_slug, version = self.default_pin(surface)
        return self.get(provider_slug, model_slug, version)

    def resolve_user(
        self, provider_slug: str | None, model_slug: str | None, surface: Surface
    ) -> ModelConfig:
        if not provider_slug or not model_slug:
            raise RegistryError(f"empty preference for {surface}")
        return self._latest_enabled(provider_slug, model_slug, surface)

    def _latest_enabled(
        self, provider_slug: str, model_slug: str, surface: Surface
    ) -> ModelConfig:
        with self._lock:
            best: ModelConfig | None = None
            for spec in self._by_pin.values():
                if (
                    spec.provider_slug == provider_slug
                    and spec.model_slug == model_slug
                    and spec.enabled
                    and spec.allows(surface)
                    and (best is None or spec.version > best.version)
                ):
                    best = spec
        if best is not None:
            return best
        with db.connect() as conn, conn.cursor() as cur:
            cur.execute(
                _select_cols
                + """
                 WHERE provider_slug=%s AND model_slug=%s
                   AND enabled AND %s = ANY(surfaces)
                 ORDER BY version DESC LIMIT 1
                """,
                (provider_slug, model_slug, surface),
            )
            row = cur.fetchone()
        if not row:
            raise RegistryError(
                f"no enabled {provider_slug}/{model_slug} for {surface}"
            )
        spec = _from_row(row)
        with self._lock:
            self._by_pin[spec.pin] = spec
        return spec


def _cached_rate_from_row(row: tuple[Any, ...]) -> int:
    if len(row) <= 16 or row[16] is None:
        raise RegistryError(
            f"{row[3]}/{row[4]} v{row[0]} is missing micros_per_cached_input_token"
        )
    return int(row[16])


def _as_dict(raw: Any) -> dict[str, Any]:
    return raw if isinstance(raw, dict) else {}


def _from_row(row: tuple[Any, ...]) -> ModelConfig:
    params = _as_dict(row[7])
    surfaces = tuple(Surface(value) for value in (row[8] or ()))
    thinking = tuple(row[9] or ())
    defaults = tuple(Surface(value) for value in (row[14] or ()))
    window = int(row[15]) if row[15] is not None else 0
    return ModelConfig(
        version=int(row[0]),
        provider_name=str(row[1] or ""),
        model_name=str(row[2] or ""),
        provider_slug=row[3],
        model_slug=str(row[4] or ""),
        platform_enabled=bool(row[5]),
        byok_enabled=bool(row[6]),
        params=params,
        surfaces=surfaces,
        thinking_levels=thinking,
        default_thinking=str(row[10] or ""),
        micros_per_input_token=int(row[11]),
        micros_per_output_token=int(row[12]),
        enabled=bool(row[13]),
        is_default_for=defaults,
        context_window_tokens=window,
        micros_per_cached_input_token=_cached_rate_from_row(row),
    )


registry = Registry()


def resolve_pinned(
    provider_slug: str | None,
    model_slug: str | None,
    version: int | None,
    surface: Surface,
) -> ModelConfig:
    """Load an exact pin, for every surface without exception.

    A missing or unresolvable pin is an error, never the live default. Whoever
    is downstream of this call is about to spend money against a price somebody
    else already quoted (the gateway's reservation, the job payload) or to write
    a vector into a space somebody else already chose (the workspace). Resolving
    a default here would run a different model than the one that was priced, and
    nothing would say so.
    """
    if not provider_slug or not model_slug or not version:
        raise RegistryError(f"missing pin for {surface}")
    return registry.get(provider_slug, model_slug, int(version))


def ingest_spec() -> ModelConfig:
    """The ingest LLM this job was enqueued with."""
    pins = current_job_pins()
    if pins is None or pins.ingest is None:
        raise RegistryError("no ingest pin on this job")
    return pins.ingest


def embedding_spec() -> ModelConfig:
    """The embedding model of the workspace this job belongs to.

    Installed by the worker from the workspace embedding pin before any
    indexing runs. Query-time embedding has no job and therefore does not come
    through here — see ``retrieval.search``.
    """
    pins = current_job_pins()
    if pins is None or pins.embedding is None:
        raise RegistryError("no embedding pin on this job")
    return pins.embedding


def vision_spec() -> ModelConfig:
    """The vision model this job was enqueued with."""
    pins = current_job_pins()
    if pins is None or pins.vision is None:
        raise RegistryError("no vision pin on this job")
    return pins.vision


def provider_api_key_for(spec: ModelConfig) -> str:
    """Return the key elitellm should use for this pin.

    User keys are passed per request and never written to process env.
    Platform keys come from the elitellm provider env name.
    """
    from .elitellm.providers import platform_api_key as elitellm_platform_key
    from .elitellm.providers import platform_env_name

    req = current_request_llm()
    if req.paid_by == "user":
        key = req.user_api_key or _load_user_key(req.user_id, spec.provider_slug)
        if not key or not spec.byok_enabled:
            raise RegistryError(
                f"user key required for {spec.provider_slug}/{spec.model_slug}"
            )
        return key
    if req.user_api_key and spec.byok_enabled and not spec.platform_enabled:
        return req.user_api_key
    if spec.byok_enabled and not spec.platform_enabled:
        raise RegistryError(
            f"user key required for {spec.provider_slug}/{spec.model_slug}"
        )
    try:
        env_name = platform_env_name(spec.provider_slug)
        key = elitellm_platform_key(spec.provider_slug)
    except KeyError as exc:
        raise RegistryError(f"unknown elitellm provider {spec.provider_slug}") from exc
    if not key:
        raise RegistryError(
            f"missing {env_name} for {spec.provider_slug}/{spec.model_slug}"
        )
    return key


def _load_user_key(user_id: str, provider_slug: str) -> str:
    if not user_id or not provider_slug:
        return ""
    from . import credentials

    key = credentials.decrypt_user_provider_key(user_id, provider_slug)
    req = current_request_llm()
    req.user_api_key = key
    return key


def context_window(spec: ModelConfig) -> int:
    if spec.context_window_tokens <= 0:
        raise RegistryError(
            f"model {spec.provider_slug}/{spec.model_slug} requires a positive "
            "context window"
        )
    return spec.context_window_tokens


def input_budget(spec: ModelConfig) -> int:
    return max(4000, context_window(spec) - 8192)


def credits_for_tokens(
    spec: ModelConfig,
    kind: str,
    input_tokens: int,
    output_tokens: int,
    cached_read_tokens: int = 0,
) -> int:
    cached = cached_read_tokens
    if cached < 0 or cached > input_tokens:
        cached = 0
    uncached = input_tokens - cached
    if kind == "embedding":
        return (
            input_tokens * spec.micros_per_input_token
            + output_tokens * spec.micros_per_input_token
        )
    return (
        uncached * spec.micros_per_input_token
        + cached * spec.micros_per_cached_input_token
        + output_tokens * spec.micros_per_output_token
    )


def poll_forever() -> None:
    while True:
        time.sleep(POLL_INTERVAL)
        try:
            registry.refresh()
        except Exception:
            log.exception("model registry poll failed")


def pins_from_payload(payload: dict[str, Any], *, embedding: ModelConfig) -> JobPins:
    """Resolve the models one ingest job is allowed to use.

    The ingest LLM and the vision model are snapshotted onto the payload at
    enqueue, because their surface defaults are hot-reloadable and the job may
    run long after the upload returned. ``embedding`` is not on the payload: it
    belongs to the workspace for the lifetime of that workspace, so the worker
    passes in the pin it read from the workspace row.

    Raises rather than returning a partial set. A job that cannot say which
    model it was priced for must not run.
    """

    def load(prefix: str, surface: Surface) -> ModelConfig:
        return resolve_pinned(
            payload.get(f"{prefix}ProviderSlug"),
            payload.get(f"{prefix}ModelSlug"),
            payload.get(f"{prefix}ModelVersion"),
            surface,
        )

    return JobPins(
        ingest=load("ingest", Surface.INGEST),
        embedding=embedding,
        vision=load("vision", Surface.VISION),
    )

File: server/internal/models/registry.go
// Package models is the hot-reloadable model registry.
//
// Rows in model_configs are immutable and versioned. The cache therefore never
// evicts a (provider_slug, model_slug, version) pin it has loaded. Polling model_registry_state
// only teaches a replica the *current defaults*; a pinned pair this process has
// never seen is a point read of the table, cached forever afterwards.
//
// A cache miss is never allowed to fall back to the current default: that would
// quietly reprice an in-flight pin. A miss the database also cannot satisfy is
// an error. Rows are disabled rather than deleted for that reason.
//
// Nothing here is frozen at process start. Retrieval and captioning defaults
// used to be, so that a 30s poll could not mix vector spaces within one corpus,
// but that made the model a query was embedded with depend on when the
// container last booted and left two replicas legitimately disagreeing. The
// guarantee now comes from pins instead: the embedding model belongs to the
// workspace for its lifetime, and the captioning model is snapshotted onto each
// ingest job at enqueue.
package models

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"sync"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/evonotes/server/internal/obs"
)

const PollInterval = 10 * time.Minute

const (
	AuthPlatform       = "platform"
	AuthUserKey        = "user_key"
	AuthPlatformOrUser = "platform_or_user"
	PaidByPlatform     = "platform"
	PaidByUser         = "user"
)

const modelConfigSelect = `
		SELECT version, provider_name, model_name, provider_slug, model_slug,
		       platform_enabled, byok_enabled, context_window_tokens,
		       thinking_levels, default_thinking, params, slots, capabilities,
		       micros_per_input_token, micros_per_output_token, enabled, is_default_for,
		       micros_per_cached_input_token
		  FROM model_configs`

var ErrNotFound = errors.New("model config not found")

// Ref is the natural identity of one provider model. Keep it structured rather
// than joining the slugs into an ambiguous string.
type Ref struct {
	ProviderSlug string `json:"providerSlug"`
	ModelSlug    string `json:"modelSlug"`
}

func (r Ref) Zero() bool { return r.ProviderSlug == "" || r.ModelSlug == "" }

func (r Ref) String() string { return r.ProviderSlug + "/" + r.ModelSlug }

// Pin is the provider/model/version tuple written onto a conversation or job. It is
// the only identifier a later request is allowed to resolve.
type Pin struct {
	Ref
	Version int `json:"modelVersion"`
}

func (p Pin) Zero() bool { return p.Ref.Zero() || p.Version <= 0 }

// Config is one immutable model_configs row.
type Config struct {
	Version                   int
	ProviderName              string
	ModelName                 string
	ProviderSlug              string
	ModelSlug                 string
	PlatformEnabled           bool
	ByokEnabled               bool
	ContextWindowTokens       int
	ThinkingLevels            []string
	DefaultThinking           string
	Params                    map[string]any
	Slots                     []string
	Capabilities              []string
	MicrosPerInputToken       int64
	MicrosPerOutputToken      int64
	Enabled                   bool
	IsDefaultFor              []string
	MicrosPerCachedInputToken int64
}

func (c Config) Ref() Ref { return Ref{ProviderSlug: c.ProviderSlug, ModelSlug: c.ModelSlug} }

func (c Config) Pin() Pin { return Pin{Ref: c.Ref(), Version: c.Version} }

func (c Config) DisplayName() string {
	return JoinModelLabel(c.ProviderName, c.ModelName)
}

func (c Config) Allows(slot string) bool {
	for _, s := range c.Slots {
		if s == slot {
			return true
		}
	}
	return false
}

func (c Config) DefaultFor(slot string) bool {
	for _, s := range c.IsDefaultFor {
		if s == slot {
			return true
		}
	}
	return false
}

func (c Config) Auth() string {
	switch {
	case c.PlatformEnabled && c.ByokEnabled:
		return AuthPlatformOrUser
	case c.ByokEnabled:
		return AuthUserKey
	default:
		return AuthPlatform
	}
}

// Available is whether this user may select the row. Platform rows are
// always selectable. BYOK-only rows need a credential.
func (c Config) Available(hasCred bool) bool {
	if c.PlatformEnabled {
		return true
	}
	return c.ByokEnabled && hasCred
}

// UsesUserKey is whether a request from this user should call the provider
// with their credential.
func (c Config) UsesUserKey(hasCred bool) bool {
	return c.ByokEnabled && hasCred
}

const (
	ThinkingInstant = "instant"
	ThinkingLow     = "low"
	ThinkingMid     = "mid"
	ThinkingHigh    = "high"
	ThinkingMax     = "max"
)

var knownThinkingLevels = map[string]struct{}{
	ThinkingInstant: {},
	ThinkingLow:     {},
	ThinkingMid:     {},
	ThinkingHigh:    {},
	ThinkingMax:     {},
}

func IsKnownThinking(level string) bool {
	_, ok := knownThinkingLevels[level]
	return ok
}

func (c Config) ResolveThinking(stored string) (string, error) {
	if stored == "" {
		if containsString(c.ThinkingLevels, c.DefaultThinking) {
			return c.DefaultThinking, nil
		}
		return "", fmt.Errorf("model %s has no valid default thinking", c.Ref())
	}
	if containsString(c.ThinkingLevels, stored) {
		return stored, nil
	}
	return "", fmt.Errorf("model %s does not support thinking %q", c.Ref(), stored)
}

func ValidateThinking(slots, levels []string, defaultThinking string) error {
	hasLLM := false
	hasEditor := false
	for _, slot := range slots {
		if IsLLMSlot(slot) {
			hasLLM = true
		}
		if slot == SlotEditor {
			hasEditor = true
		}
	}
	if !hasLLM {
		if len(levels) > 0 || defaultThinking != "" {
			return fmt.Errorf("retrieval/captioning rows must omit thinking")
		}
		return nil
	}
	if len(levels) == 0 {
		return fmt.Errorf("llm rows require thinking levels")
	}
	for _, item := range levels {
		if !IsKnownThinking(item) {
			return fmt.Errorf("unknown thinking level %q", item)
		}
	}
	if defaultThinking == "" || !containsString(levels, defaultThinking) {
		return fmt.Errorf("default thinking must be one of this row's levels")
	}
	if hasEditor && !containsString(levels, ThinkingInstant) {
		return fmt.Errorf("editor rows must support instant thinking")
	}
	return nil
}

func UsesResponses(providerSlug, thinking string, tools bool) bool {
	return providerSlug == "openai" && tools && thinking != "" && thinking != ThinkingInstant
}

func (c Config) UsesResponses(thinking string, tools bool) bool {
	return UsesResponses(c.ProviderSlug, thinking, tools)
}

func containsString(list []string, want string) bool {
	for _, item := range list {
		if item == want {
			return true
		}
	}
	return false
}

func (c Config) ParamFloat(key string, fallback float64) float64 {
	if c.Params == nil {
		return fallback
	}
	switch v := c.Params[key].(type) {
	case float64:
		return v
	case json.Number:
		f, err := v.Float64()
		if err == nil {
			return f
		}
	case int:
		return float64(v)
	case int64:
		return float64(v)
	}
	return fallback
}

type cacheKey struct {
	providerSlug string
	modelSlug    string
	version      int
}

func cacheKeyFor(ref Ref, version int) cacheKey {
	return cacheKey{providerSlug: ref.ProviderSlug, modelSlug: ref.ModelSlug, version: version}
}

// Registry caches immutable configs and the current defaults.
type Registry struct {
	pool *pgxpool.Pool

	mu      sync.RWMutex
	byPin   map[cacheKey]Config
	current map[string]Pin // slot -> default pin
	rev     int64
}

func New(ctx context.Context, pool *pgxpool.Pool) (*Registry, error) {
	r := &Registry{
		pool:    pool,
		byPin:   map[cacheKey]Config{},
		current: map[string]Pin{},
	}
	if err := r.refresh(ctx); err != nil {
		return nil, err
	}
	return r, nil
}

func (r *Registry) Pool() *pgxpool.Pool { return r.pool }

// Poll watches model_registry_state.version and reloads defaults when it
// changes. Retargeting a default only affects work that has not been pinned
// yet: an in-flight job, an existing conversation and an existing workspace all
// keep resolving the pair they were given.
func (r *Registry) Poll(ctx context.Context) {
	ticker := time.NewTicker(PollInterval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			if err := r.refresh(obs.Background(ctx, "model_registry")); err != nil {
				obs.CaptureErr(ctx, err, map[string]string{"stage": "model_registry_poll"})
			}
		}
	}
}

func (r *Registry) refresh(ctx context.Context) error {
	var rev int64
	if err := r.pool.QueryRow(ctx,
		`SELECT version FROM model_registry_state WHERE id = true`,
	).Scan(&rev); err != nil {
		return err
	}
	r.mu.RLock()
	same := r.rev == rev && rev != 0 && len(r.current) > 0
	r.mu.RUnlock()
	if same {
		return nil
	}

	rows, err := r.pool.Query(ctx, modelConfigSelect)
	if err != nil {
		return err
	}
	defer rows.Close()

	byPin := map[cacheKey]Config{}
	current := map[string]Pin{}
	for rows.Next() {
		cfg, err := scanConfig(rows)
		if err != nil {
			return err
		}
		byPin[cacheKeyFor(cfg.Ref(), cfg.Version)] = cfg
		if !cfg.Enabled {
			continue
		}
		for _, slot := range cfg.IsDefaultFor {
			prev, ok := current[slot]
			if !ok || cfg.Version >= prev.Version {
				current[slot] = cfg.Pin()
			}
		}
	}
	if err := rows.Err(); err != nil {
		return err
	}

	r.mu.Lock()
	defer r.mu.Unlock()
	for k, cfg := range byPin {
		r.byPin[k] = cfg
	}
	for slot, pin := range current {
		r.current[slot] = pin
	}
	r.rev = rev
	return nil
}

type rowScanner interface {
	Scan(dest ...any) error
}

func scanConfig(row rowScanner) (Config, error) {
	var (
		c            Config
		params       []byte
		slots        []string
		capabilities []string
		defaultFor   []string
		thinking     []string
	)
	err := row.Scan(
		&c.Version, &c.ProviderName, &c.ModelName, &c.ProviderSlug, &c.ModelSlug,
		&c.PlatformEnabled, &c.ByokEnabled, &c.ContextWindowTokens,
		&thinking, &c.DefaultThinking,
		&params, &slots, &capabilities, &c.MicrosPerInputToken, &c.MicrosPerOutputToken,
		&c.Enabled, &defaultFor, &c.MicrosPerCachedInputToken,
	)
	if err != nil {
		return Config{}, err
	}
	c.ThinkingLevels = thinking
	c.Slots = slots
	c.Capabilities = capabilities
	c.IsDefaultFor = defaultFor
	if len(params) > 0 {
		_ = json.Unmarshal(params, &c.Params)
	}
	if c.Params == nil {
		c.Params = map[string]any{}
	}
	return c, nil
}

// Get returns the exact provider/model/version pin. Load-on-miss from the table; never
// falls back to the current default. Disabled chat/generate rows still
// resolve so a pinned conversation keeps working. Retrieval rows cannot be
// disabled or rewritten onto a different model: Postgres rejects the write.
func (r *Registry) Get(ctx context.Context, ref Ref, version int) (Config, error) {
	if ref.Zero() || version <= 0 {
		return Config{}, ErrNotFound
	}
	ck := cacheKeyFor(ref, version)
	r.mu.RLock()
	if cfg, ok := r.byPin[ck]; ok {
		r.mu.RUnlock()
		return cfg, nil
	}
	r.mu.RUnlock()

	cfg, err := r.load(ctx, ref, version)
	if err != nil {
		return Config{}, err
	}
	r.mu.Lock()
	r.byPin[ck] = cfg
	r.mu.Unlock()
	return cfg, nil
}

func (r *Registry) load(ctx context.Context, ref Ref, version int) (Config, error) {
	row := r.pool.QueryRow(ctx, modelConfigSelect+`
		 WHERE provider_slug=$1 AND model_slug=$2 AND version=$3`, ref.ProviderSlug, ref.ModelSlug, version)
	cfg, err := scanConfig(row)
	if errors.Is(err, pgx.ErrNoRows) {
		return Config{}, fmt.Errorf("%w: %s v%d", ErrNotFound, ref, version)
	}
	return cfg, err
}

// DefaultPin is the current default for a slot. Callers that are choosing on
// somebody's behalf (enqueue, account creation, workspace creation) resolve it
// once and store the result; nothing downstream is allowed to call this again.
func (r *Registry) DefaultPin(slot string) (Pin, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	pin, ok := r.current[slot]
	if !ok || pin.Zero() {
		return Pin{}, fmt.Errorf("%w: no default for %s", ErrNotFound, slot)
	}
	return pin, nil
}

func (r *Registry) Default(ctx context.Context, slot string) (Config, error) {
	pin, err := r.DefaultPin(slot)
	if err != nil {
		return Config{}, err
	}
	return r.Get(ctx, pin.Ref, pin.Version)
}

// ResolveUser returns the latest enabled config for the user's preferred model
// on this slot. The preference must be a non-empty ref that still resolves;
// there is no fallback to the slot default. Account creation snapshots the
// default onto the user row so a live request always has a concrete model ref.
func (r *Registry) ResolveUser(ctx context.Context, pref Ref, slot string) (Config, error) {
	if pref.Zero() {
		return Config{}, fmt.Errorf("%w: empty preference for %s", ErrNotFound, slot)
	}
	return r.latestEnabled(ctx, pref, slot)
}

func (r *Registry) latestEnabled(ctx context.Context, ref Ref, slot string) (Config, error) {
	r.mu.RLock()
	var best Config
	found := false
	for _, cfg := range r.byPin {
		if cfg.Ref() == ref && cfg.Enabled && cfg.Allows(slot) {
			if !found || cfg.Version > best.Version {
				best = cfg
				found = true
			}
		}
	}
	r.mu.RUnlock()
	if found {
		return best, nil
	}
	row := r.pool.QueryRow(ctx, modelConfigSelect+`
		 WHERE provider_slug=$1 AND model_slug=$2 AND enabled AND $3 = ANY(slots)
		 ORDER BY version DESC LIMIT 1`, ref.ProviderSlug, ref.ModelSlug, slot)
	cfg, err := scanConfig(row)
	if errors.Is(err, pgx.ErrNoRows) {
		return Config{}, fmt.Errorf("%w: %s", ErrNotFound, ref)
	}
	if err != nil {
		return Config{}, err
	}
	r.mu.Lock()
	r.byPin[cacheKeyFor(cfg.Ref(), cfg.Version)] = cfg
	r.mu.Unlock()
	return cfg, nil
}

// SnapshotIngest returns the pins written onto an ingest job at enqueue: the
// two slots whose defaults are hot-reloadable and whose job may still be
// queued when one moves.
//
// Retrieval is deliberately absent. It comes from the workspace row instead,
// because a workspace's vector space outlives any single job and must not be
// re-decided per upload.
func (r *Registry) SnapshotIngest(ctx context.Context) (ingest, captioning Config, err error) {
	ingest, err = r.Default(ctx, SlotIngest)
	if err != nil {
		return
	}
	captioning, err = r.Default(ctx, SlotCaptioning)
	return
}

// EmbeddingDim is the vector width a config emits. The vector table is chosen
// by pin, not by width; this value sizes the halfvec write. A check constraint
// in the migration requires it on every retrieval row, so a zero here means
// the row was written around the schema.
func (c Config) EmbeddingDim() (int, error) {
	dim := int(c.ParamFloat("dimensions", 0))
	if dim <= 0 {
		return 0, fmt.Errorf("%w: %s v%d declares no dimensions", ErrNotFound, c.Ref(), c.Version)
	}
	return dim, nil
}

// ListEnabled returns enabled configs that advertise slot, newest version
// per provider/model identity, for the model picker.
func (r *Registry) ListEnabled(slot string) []Config {
	r.mu.RLock()
	defer r.mu.RUnlock()
	best := map[Ref]Config{}
	for _, cfg := range r.byPin {
		if !cfg.Enabled || !cfg.Allows(slot) {
			continue
		}
		ref := cfg.Ref()
		prev, ok := best[ref]
		if !ok || cfg.Version > prev.Version {
			best[ref] = cfg
		}
	}
	out := make([]Config, 0, len(best))
	for _, cfg := range best {
		out = append(out, cfg)
	}
	return out
}

func (r *Registry) Rev() int64 {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.rev
}
